MM-27212: Cypress/e2e: Add user to channel from popover

agnivade opened the pull request:
This PR automates
- MM-T3
- MM-T4
- MM-T5
- MM-T6
- MM-T7
- MM-T8
- MM-T9
- MM-T1

https://mattermost.atlassian.net/browse/MM-27212

<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
<!--
A description of what this pull request does.
-->

#### Ticket Link
<!--
If this pull request addresses a Help Wanted ticket, please link the relevant GitHub issue, e.g.

  Fixes https://github.com/mattermost/mattermost-server/issues/XXXXX

Otherwise, link the JIRA ticket.
-->

agnivade commented on e2e/cypress/integration/add_user_to_channel_from_profile_popover/profile_popover_spec.js:
```diff
@@ -31,29 +37,12 @@ describe('Profile popover', () => {
         // # Do the system console setup
         cy.apiAdminLogin();
         cy.visit('/admin_console/user_management/permissions/system_scheme');
-        cy.findByTestId('all_users-public_channel-checkbox').scrollIntoView().should('be.visible').click();
```
@saturninoabril - you were right. Clicking it did disable it. Removed it now. Not sure why it wasn't working yesterday.

saturninoabril commented:
Thanks @agnivade! Could you please check again if you're getting consistent passing results? I've got several failing tests running the spec.
<img width="501" alt="Screen Shot 2020-09-08 at 7 23 26 PM" src="https://user-images.githubusercontent.com/5334504/92470447-00c8ab80-f209-11ea-955a-bf31fe6d226c.png">

agnivade commented:
@saturninoabril - Can you tell me what is failing on T4? I think it's a cascading failure from there. I cannot seem to repro it here.

saturninoabril commented:
It failed on checking "Manage Public Channels".
![Screen Shot 2020-09-08 at 10 30 06 PM](https://user-images.githubusercontent.com/5334504/92490647-db956680-f223-11ea-8eb4-ae49955794bb.png)

- Do you need license to run each test? If some require license, then I suggest to group and move into a separate spec file with `cy.apiRequireLicense()` check at before hook.
- Each test seemed to be dependent with previous test. I suggest to run the test with `it.only` and see if it'll work as expected. You might also need to reset before each test. Below might help,
```javascript
beforeEach(() => {
    cy.apiAdminLogin();
    cy.apiResetRoles();
});
```

agnivade commented:
> Each test seemed to be dependent with previous test.

Yep, that needs to be fixed. I'll look into it.

agnivade commented:
@saturninoabril - should be much better now. PTAL.

streamer45 commented on e2e/cypress/integration/add_user_to_channel_from_profile_popover/profile_popover_spec.js:
```diff
@@ -88,4 +77,465 @@ describe('Profile popover', () => {
+        classCondition = 'not.have.class';
+    }
+
+    // # Expand the section
+    cy.get('#' + category + '-' + publicOrPrivate + '_channel > .fa').scrollIntoView().should('be.visible').click();
+
+    if (category !== 'channel_admin') {
+        cy.findByTestId(category + '-' + publicOrPrivate + '_channel-create_' + publicOrPrivate + '_channel-checkbox').should(classCondition, 'checked');
+    }
+    cy.findByTestId(category + '-' + publicOrPrivate + '_channel-manage_' + publicOrPrivate + '_channel_properties-checkbox').should(classCondition, 'checked');
+    cy.findByTestId(category + '-' + publicOrPrivate + '_channel-manage_' + publicOrPrivate + '_channel_members_and_read_groups-checkbox').should(classCondition, 'checked');
+    cy.findByTestId(category + '-' + publicOrPrivate + '_channel-delete_' + publicOrPrivate + '_channel-checkbox').should(classCondition, 'checked');
+};
```
Using [template literals](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals) (string in backticks) might improve the readability a bit.

streamer45 reviewed (approved):
This was a lot of work. Nicely done :+1: 
Left one small suggestion.

saturninoabril commented:
@agnivade I'm still seeing consistent failing tests.
<img width="515" alt="Screen Shot 2020-09-09 at 6 30 08 PM" src="https://user-images.githubusercontent.com/5334504/92587681-8eb89b00-f2ca-11ea-89cc-f99f627963be.png">

agnivade commented:
Woops, this is bad. I think somehow my environment gets polluted with other things. I'll dig some more.

agnivade commented:
Hey @saturninoabril - I tried multiple runs with a fresh DB and it passes every time. Are you getting the same error on T4 as you saw earlier here https://github.com/mattermost/mattermost-webapp/pull/6395#issuecomment-688925963?

saturninoabril commented:
Hmm, not sure why it turned out like that but here's the latest run on me.
<img width="515" alt="Screen Shot 2020-09-09 at 6 30 08 PM" src="https://user-images.githubusercontent.com/5334504/92627960-b1fd3d80-f2fe-11ea-973b-a81e16783f47.png">

saturninoabril commented on e2e/cypress/integration/add_user_to_channel_from_profile_popover/profile_popover_spec.js:
```diff
@@ -88,4 +77,465 @@ describe('Profile popover', () => {
+        // # Do the system console setup
+        cy.apiAdminLogin();
+        cy.visit('/admin_console/user_management/permissions/system_scheme');
+        cy.findByTestId('all_users-public_channel-checkbox').scrollIntoView().should('be.visible').click();
```
Looks like this one requires another click to fully uncheck the "Manage Public Channels". Initial state has "Convert Channels" uncheck on me and so first click makes the entire "Manage Public Channels" checked.

agnivade replied:
Ah looks like that was it. I had an outdated copy of the webapp. Fixing.

saturninoabril commented on e2e/cypress/integration/add_user_to_channel_from_profile_popover/profile_popover_spec.js:
```diff
@@ -88,4 +77,465 @@ describe('Profile popover', () => {
     });
+
+    it('MM-T3 Add user — Public ON / Private OFF', () => {
+        // # Do the system console setup
+        cy.apiAdminLogin();
+        cy.visit('/admin_console/user_management/permissions/system_scheme');
```
I see this repeated in several test blocks. Maybe move to `beforeEach`? Other that don't require it might be better to move in a separate spec file.

agnivade replied:
Hmm .. it might look a bit weird to have T5 and T8 in a separate file and the rest in another, just to de-duplicate 2 lines. I think this is fine as is. And groups all related tests in a single file. 0/5.

saturninoabril replied:
I'll defer to you but just saying for test maintainability where we spend most of the time after the script is merged. Also, on the same note, making each test isolated will help us easily debug the script in the future. You'll know it's isolated if you can run the test successfully after adding `it.only` on one `it` block.

agnivade replied:
No worries, I'll move it out. The tests are not 100% isolated because the posts created are shared between them but the user roles being reset and now some more tests being moved out should make them stable.

saturninoabril commented on e2e/cypress/integration/add_user_to_channel_from_profile_popover/profile_popover_spec.js:
```diff
@@ -88,4 +77,465 @@ describe('Profile popover', () => {
+        // # Do the system console setup
+        cy.apiAdminLogin();
+        cy.visit('/admin_console/user_management/permissions/system_scheme');
+        cy.findByTestId('all_users-public_channel-checkbox').scrollIntoView().should('be.visible').click();
```
Same here for another click to fully uncheck.

agnivade replied:
done

saturninoabril commented on e2e/cypress/integration/add_user_to_channel_from_profile_popover/profile_popover_spec.js:
```diff
@@ -88,4 +77,465 @@ describe('Profile popover', () => {
+        // # Do the system console setup
+        cy.apiAdminLogin();
+        cy.visit('/admin_console/user_management/permissions/system_scheme');
+        cy.findByTestId('all_users-public_channel-checkbox').scrollIntoView().should('be.visible').click();
```
And here.

agnivade replied:
done

saturninoabril reviewed (approved):
Thanks @agnivade for updating. LGTM, tested and all passed.

agnivade commented:
@saturninoabril - I split them into different spec files now and also added check for the convert channel checkbox. Could you please run the tests once more from your end to confirm everything is okay? Thanks.

saturninoabril commented:
Thanks @agnivade. Tested and all passed.

lindy65 reviewed (approved):
Thanks @agnivade - looks good to me 👍

mattermod commented:
Will try to auto merge this PR once all tests and checks are passing. This might take up to an hour.

mattermod commented:
Trying to auto merge this PR.

mattermod commented:
Pull Request successfully merged
SHA: 390585f9761a72041b73389f49d12c5ecad8b9ce
